Add RestTestClient

rworsnop opened the pull request:
Fixes gh-31275

rworsnop committed:
Add RestTestClient

Fixes gh-31275

Signed-off-by: Rob Worsnop <rworsnop@gmail.com>

rstoyanchev reviewed (changes requested):
Thanks for the detailed contribution. 

Generally, seems okay on a first pass. Some initial comments on packaging and naming. Let me know if you would you like to continue to iterate over the PR based on my feedback, or if you would prefer me to take over?

I see the symmetry between `~.web.server` package for `~RestTestClient` and `~.web.reactive.server` for `WebTestClient`, but `~.web.servlet` is an already established home for server tests with Spring MVC, and having a new package next to that creates an overlap.  So, `~.web.servlet` is the "server" package, and this addition can live somewhere undertneath it as a layer over MockMvc alongside others like `htmlunit` and `assertj`.

It's true that `RestTestClient` is for both MockMvc _and_ for live server testing, but the same applies to `WebTestClient` as well, which is in one package, and that's fine. It's basically a test client for MockMvc with the extra option to call a live server (anything that implements `ClientHttpRequestsFactory`).

We have `~.web.servlet.client` where `MockMvcWebTestClient` lives. That package is ideal, and rather than thinking of a new name, we can put `RestTestClient` support there as the eventual successor of `MockMvcWebTestClient`. In the long run there is no need for both to remain so that should eventually clear up.

This would create a conflict with new and existing `AbstractMockMvcServerSpec` and `ApplicationContextMockMvcSpec` classes, but it is already confusing to have those IMO, both for MockMvc, but in different packages. Instead if we drop "Mvc" from the name in the new classes, and essentially follow the same convention as in `WebTestClient`, i.e. `AbstractMockServerSpec` and `ApplicationContextSpec`, that would give us two classes with the same name, but one for WebFlux and one for MockMvc, which would be fine.

In short my first suggstion is to move `RestTestClient` into `.web.servlet.client`, and rename server spec classes accordingly. 

In addition I suggest to create a copy of `MockMvcClientHttpRequestFactory` in the same package as `RestTestClient`, and deprecate the one currently in `~.web.client`. It was added there all the way back in 3.2 because there was no better place at the time, but it has nothing to do with the rest of what's in that package, and makes sense to have it in the same package as `RestTestClient`. For anyone currently using that class, this would serve as an alert that there is now higher level support and most likely no need to use that class directly. That could be described in the deprecation notice.

rworsnop commented:
Thanks for the feedback. I will work on this next week.

rworsnop committed:
Rename AbstractMockServerSpec and friends

Fixes gh-31275

Signed-off-by: Rob Worsnop <rworsnop@gmail.com>

rworsnop committed:
Move web.server classes into web.servlet.client package

Fixes gh-31275

Signed-off-by: Rob Worsnop <rworsnop@gmail.com>

rworsnop committed:
Deprecate MockMvcClientHttpRequestFactory and copy into servlet.client

Fixes gh-31275

Signed-off-by: Rob Worsnop <rworsnop@gmail.com>

rstoyanchev reviewed (approved):
That looks good overall now. Thanks for making those changes. I will have a more detailed look. No worries about the commits, I will squash accordingly before merging.

rstoyanchev reviewed (changes requested):
Please, see some feedback on the API to create `RestTestClient` instances. Let me know what you think, and if you intend to make the changes.

rworsnop committed:
Expose MockMvc in API

Fixes gh-31275

Signed-off-by: Rob Worsnop <rworsnop@gmail.com>

rworsnop commented:
Done. Always nice to delete code!

rworsnop committed:
Refactor *Assertions classes

Fixes gh-31275

Signed-off-by: Rob Worsnop <rworsnop@gmail.com>

rworsnop commented:
I just added another commit. 

The *Assertions classes are very similar to the reactive ones, so I moved the public methods into a new abstract class for `CookieAssertions`.

What do you think? If this is considered beyond the scope for this PR I can roll back, otherwise I can do the same with the other *Assertions classes as well.

rstoyanchev commented:
Good call on the *Assertions classes. There are quite extensive and detailed if not complicated. I don't think they should be in the web package though. I would suggest a `~.web.support~` package for that.

rstoyanchev commented:
I did some more iterations where the feedback would have been too detailed. It's broken out into logical commits so hopefully straight forward to follow. 

The big one was re-ordering the internal structure to make `RestTestClient` and `WebTestClient` sources directly comparable, which helps to ensure the two are aligned in terms of feature set now and also over time. I also made further updates to the Builder hierarchy to ensure user-facing type names are simpler, without generics, and that the concrete builder type carries forward with method chaining.

If there are any questions, please reach out. Thanks for this significant contribution!

spencergibb commented:
So happy this got merged https://github.com/spring-cloud/spring-cloud-gateway/pull/3872

rworsnop commented:
Yay! First customer!
